### ISSUE TITLE
Open authOrigins in a SFSafariView by default?

### ISSUE THREAD

**Issue by yocontra** [pwa-builder/pwabuilder-ios#30]
Currently Google OAuth throws an error if you try to load it in a WebView - this is by design, they don't want people using in-app web views for OAuth flows. If you open it in a SFSafariView it works fine, because this is an isolated sandbox.

My proposal is to change the behavior of the `authOrigins` default to use SFSafariView instead of opening it full-screen (logic is already there for opening http links) - I think it achieves the same basic behavior but allows it to work with google oauth.

Another issue with the `authOrigins` behavior - it should take precedence over the `requestHost = requestUrl.host` conditional. In many cases (passport being one) you have a specific route on your webserver that redirects you to the signed OAuth URL. In a case like this you would want to add something like `authOrigins = [ "accounts.google.com", "mybackend.com/auth/google/start" ]` and expect it to open in a SFSafariView.

Using passport the flow would work like so:
- App opens `mybackend.com/auth/google/start` which loads in a SFSafariView (user specified this URL in `authOrigins`)
- `mybackend.com/auth/google/start` redirects to signed `accounts.google.com` URL
- User authenticates with google, google redirects to `mybackend.com/auth/google/some_callback_url` which registers the session on the backend, sets a cookie, and redirects to `mybackend.com`
- Because app links are set up, this redirect to `mybackend.com` should kill the open SFSafariView and navigate to the URL (this would be an additional behavior change)

So tl;dr I think with some changes to deep link handling (kill any open external web views), tweaks to authOrigins logic, and landing #28 you could have passport.js be pretty close to drop-in compatibility. I'm not sure if #28 allows you to share credentials between a SFSafariView and a WebView, but that is really a nice to have that could be worked around (the back-end could pass the session token via the deep link instead of using cookies).

**Comment by khmyznikov** [comment]
SFSafariView doesn't share anything with WKWebView, so I don't see a possibility to change this logic. You should use the redirect-login process with bound domains or you just can't get anything back. Backend workarounds are not for all.

**Comment by khmyznikov** [comment]
You can freely change this logic for your app, it's just a template first of all. I'm sure we can't cover here all possible variants of apps.

**Comment by yocontra** [comment]
@khmyznikov Yep understood that it is just a template - I think the defaults could be improved which is why I opened the issue. I think I proposed more changes here than just SFSafariView being the default, but if you think all of these are out of scope feel free to close this. A note in the docs that Google OAuth does not work (they simply block the user-agent for a iOS WebView) would probably be appreciated by some so they go down this same rabbit hole.

For context we're currently using capacitor and tried to port our app over to use this - worked great for android (0 changes needed, just worked out of the box), and not at all for iOS due to the OAuth redirect flow issues outlined above. Won't be super upset or anything if we have to stick with capacitor for the iOS app, but I felt like this was worthy to report + open for discussion.

**Comment by yocontra** [comment]
Also FWIW I don't think this backend flow I outlined is a workaround - this is how most apps are handling OAuth today, I singled out passport.js specifically as that is what we use but other frameworks operate similarly. What I was trying to propose was a way for pwabuilder to support these existing OAuth flows without requiring any code changes.

**Comment by yocontra** [comment]
This may not be interesting to you but another way to fix this is you can just override the user-agent of the webview to imitate a normal SFSafariView so it doesn't get blocked ^ kind of a hacky workaround though

**Comment by khmyznikov** [comment]
@yocontra can you specify how user-agent looks for webview and SF?

**Comment by yocontra** [comment]
@khmyznikov More info on it here: https://lists.webkit.org/pipermail/webkit-dev/2020-November/031604.html and discussion here: https://news.ycombinator.com/item?id=25155451


I think by default an in-app webview has no user-agent which is how they are detecting/blocking it, so something as simple as this also solves the issue:

```
webView.getSettings().setUserAgentString(System.getProperty("http.agent"))
```

This uses the system safari user-agent for the webview, which is the same as SFSafariView

**Comment by justrealmilk** [comment]
Would changing web view controls also fix the fact that there's no URL or connection type (friendly SSL padlock) in the title bar? Users like that assurance for OAuth

**Comment by khmyznikov** [comment]
You can't change all auth methods by default to SF because if something use cookies or storage it's couldn't be passed back.

**Comment by justrealmilk** [comment]
> @khmyznikov More info on it here: https://lists.webkit.org/pipermail/webkit-dev/2020-November/031604.html and discussion here: https://news.ycombinator.com/item?id=25155451
> 
> I think by default an in-app webview has no user-agent which is how they are detecting/blocking it, so something as simple as this also solves the issue:
> 
> ```
> webView.getSettings().setUserAgentString(System.getProperty("http.agent"))
> ```
> 
> This uses the system safari user-agent for the webview, which is the same as SFSafariView. This solves the issue but doesn't really address the security concerns of why Google is telling people to use SFSafariView - with a WebView you can't tell if you are being phished or not, SFSafariView you can see the address and the parent app has no access to cookie values or other internals.

I assumed/hoped that was valid Swift code. Alas.

I used `webView.customUserAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Mobile/15E148 Safari/604.1"` because I couldn't work out how to grab Safari's agent at runtime. 🤫

**Comment by kyleshay** [comment]
FWIW This also does the the trick in `WebView.swift#L27` by just appending the string to the existing UA:

 `config.applicationNameForUserAgent = "Safari/604.1"`

**Comment by JudahGabriel** [comment]
I'm going to add a user agent by default to the template.

**Comment by JudahGabriel** [comment]
I've updated the template to include the following lines:

```swift
webView.configuration.applicationNameForUserAgent = "Safari/604.1"
webView.customUserAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Mobile/15E148 Safari/604.1";
```

**Comment by kyleshay** [comment]
Hmm the second `customUserAgent` may not be necessary? At least it wasn't needed with just setting the `webView.configuration.applicationNameForUserAgent`. This way I'm able to preserve the existing UA from a device, for splitting by devices in analytics etc.

**Comment by yocontra** [comment]
@JudahGabriel I think that resolves this ticket - will close and reopen if anyone says otherwise. To future-proof that you may want to make it something like `WKWebView().value(forKey: "userAgent")` (found this somewhere else, YMMV)

**Comment by goyavo** [comment]
> Hmm the second `customUserAgent` may not be necessary? At least it wasn't needed with just setting the `webView.configuration.applicationNameForUserAgent`. This way I'm able to preserve the existing UA from a device, for splitting by devices in analytics etc.

+1 - we also use UA for analytics
